File: DTCore5.0-exp/DTCore/DataTools.Memory/SwapStream/VirtualMemoryStream.vb
Imports System.IO
Imports System.Runtime.InteropServices

Namespace Memory

    Public NotInheritable Class VirtualMemoryStream
        Inherits Stream

        Private _Blob As New Blob With {.InBufferMode = True, .BufferExtend = 65536, .MemoryType = MemAllocType.Virtual}

        Public Overrides ReadOnly Property CanRead As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property CanSeek As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property CanWrite As Boolean
            Get
                Return True
            End Get
        End Property

        Public Overrides ReadOnly Property Length As Long
            Get
                Return _Blob.Length
            End Get
        End Property

        Public Overrides Property Position As Long
            Get
                Return _Blob.ClipNext
            End Get
            Set(value As Long)
                _Blob.ClipSeek(value)
            End Set
        End Property

        Public Overrides Sub Flush()
            Return
        End Sub

        Public Overrides Sub SetLength(value As Long)
            _Blob.Length = value
        End Sub

        Public Overrides Sub Write(buffer() As Byte, offset As Integer, count As Integer)

            Dim gch As GCHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned)
            Dim cptr As IntPtr = gch.AddrOfPinnedObject + offset

            If (_Blob.Length - _Blob.ClipNext) < count Then
                _Blob.Length += (count - _Blob.ClipNext)
            End If

            DataTools.Memory.Internal.Native.MemCpy(_Blob.DangerousGetHandle() + _Blob.ClipNext, cptr, count)
            gch.Free()

            _Blob.ClipSeek(_Blob.ClipNext + count)

        End Sub

        Public Overrides Function Read(buffer() As Byte, offset As Integer, count As Integer) As Integer

            Dim gch As GCHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned)
            Dim cptr As IntPtr = gch.AddrOfPinnedObject + offset

            If (_Blob.Length - _Blob.ClipNext) < count Then
                count = _Blob.Length - _Blob.ClipNext
            End If

            If count <= 0 Then Return 0

            Internal.Native.MemCpy(_Blob.DangerousGetHandle() + _Blob.ClipNext, cptr, count)
            _Blob.ClipSeek(_Blob.ClipNext + count)

            Return count
        End Function

        Public Overrides Function Seek(offset As Long, origin As SeekOrigin) As Long

            Select Case origin
                Case SeekOrigin.Begin
                    _Blob.ClipSeek(offset)

                Case SeekOrigin.Current
                    _Blob.ClipSeek(_Blob.ClipNext + offset)

                Case SeekOrigin.End
                    _Blob.ClipSeek(_Blob.Length + offset)

            End Select

            Return _Blob.ClipNext
        End Function

        Protected Overrides Sub Dispose(disposing As Boolean)
            MyBase.Dispose(disposing)

            If disposing Then
                _Blob.Free()
            End If

        End Sub

    End Class



End Namespace

File: DTCore5.0-exp/DTCore/DataTools.Memory/SwapStream/VirtualMemoryStream.cs
using System.IO;
using System.Runtime.InteropServices;

namespace DataTools.Memory
{
    public sealed class VirtualMemoryStream : Stream
    {
        public VirtualMemoryStream()
        {
            _Blob = new Blob() { InBufferMode = true, BufferExtend = 65536L, MemoryType = MemAllocType.Virtual };
        }

        private Blob _Blob;

        public override bool CanRead
        {
            get
            {
                return true;
            }
        }

        public override bool CanSeek
        {
            get
            {
                return true;
            }
        }

        public override bool CanWrite
        {
            get
            {
                return true;
            }
        }

        public override long Length
        {
            get
            {
                return _Blob.Length;
            }
        }

        public override long Position
        {
            get
            {
                return _Blob.ClipNext;
            }

            set
            {
                _Blob.ClipSeek(value);
            }
        }

        public override void Flush()
        {
            return;
        }

        public override void SetLength(long value)
        {
            _Blob.Length = value;
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            var gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);
            var cptr = gch.AddrOfPinnedObject() + offset;
            if (_Blob.Length - _Blob.ClipNext < count)
            {
                _Blob.Length += count - _Blob.ClipNext;
            }

            Internal.Native.MemCpy(_Blob.DangerousGetHandle() + _Blob.ClipNext, cptr, (uint)count);
            gch.Free();
            _Blob.ClipSeek(_Blob.ClipNext + count);
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            var gch = GCHandle.Alloc(buffer, GCHandleType.Pinned);
            var cptr = gch.AddrOfPinnedObject() + offset;
            if (_Blob.Length - _Blob.ClipNext < count)
            {
                count = (int)(_Blob.Length - _Blob.ClipNext);
            }

            if (count <= 0)
                return 0;
            Internal.Native.MemCpy(_Blob.DangerousGetHandle() + _Blob.ClipNext, cptr, (uint)count);
            _Blob.ClipSeek(_Blob.ClipNext + count);
            return count;
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            switch (origin)
            {
                case SeekOrigin.Begin:
                    {
                        _Blob.ClipSeek(offset);
                        break;
                    }

                case SeekOrigin.Current:
                    {
                        _Blob.ClipSeek(_Blob.ClipNext + offset);
                        break;
                    }

                case SeekOrigin.End:
                    {
                        _Blob.ClipSeek(_Blob.Length + offset);
                        break;
                    }
            }

            return _Blob.ClipNext;
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);
            if (disposing)
            {
                _Blob.Free();
            }
        }
    }
}